classes/db/DAO.inc.php

```diff
@@ -103,8 +103,15 @@ function retrieveRange($sql, $params = [], $dbResultRange = null, $callHooks = t
 	 * @return int
 	 */
 	public function countRecords($sql, $params = []) {
-		$result = $this->retrieve('SELECT COUNT(*) AS row_count FROM (' . $sql . ') AS count_subquery', $params);
-		return $result->current()->row_count;
+		foreach ([$this->_optimizeCountQuery($sql, $params), [$sql, $params]] as [$sql, $params]) {
```

Comment:
Considering that a `COUNT()` don't need to know anything about costly SELECT fields (`SELECT (SELECT ...) AS slow_sub_query`) nor `ORDER BY`s, I think it's safe to drop them, but MySQL doesn't seem to recognize this optimization opportunity.

So the "parser" is just dropping the SELECT and the ORDER BY clauses, taking care to not touch strings/sub-queries.

Besides skipping the `UNION`, other complex cases I decided to not cover would be:
- `SELECT field AS alias... GROUP BY alias`
- `SELECT function(or_sub_query) GROUP BY function(or_sub_query)`

Optimization failures should fall into the retry mechanism, which will use the original query.

Reply:
Hmm, this is an awful lot of magic to be adding, especially to a stable release. I did run into [an implementation of this optimization in ADODB](https://github.com/ADOdb/ADOdb/blob/abe86c12b27e428eeee1f933f42f0dcf2c0dd403/adodb-lib.inc.php#L391), but it shows signs of needing [fixes](https://github.com/ADOdb/ADOdb/issues/549) on [fixes](https://github.com/ADOdb/ADOdb/issues/870) even though it's a very mature codebase. (In my opinion this kind of thing is what has effectively killed ADODB; it's become impossible to maintain because of the potential for complicated interactions. Touch one thing and you're guaranteed to unbalance a dozen carefully-balanced other things.)

In the `main` branch, the relevant code has already been rewritten into the Laravel query builder toolset and Collector pattern (e.g. https://github.com/pkp/pkp-lib/blob/main/classes/user/Collector.php and https://github.com/pkp/pkp-lib/blob/main/classes/userGroup/Collector.php), which gives more of an ability to intervene in how SQL is built; I think this optimization to drop ordering could be applied to `main` without needing to build and then parse the entire SQL string. You could just drop the `ORDER BY` part of the query builder and get on with your day. I'd be more amenable to an approach like that.

Reply:
There's a very close parallel to what I'm proposing e.g. [here](https://github.com/pkp/pkp-lib/blob/d8244127bff8078f834f31f23ed29fee11a9a7fa/api/v1/users/PKPUserHandler.php#L142..L153), where the same QueryBuilder is first used to [get a page of results](https://github.com/pkp/pkp-lib/blob/d8244127bff8078f834f31f23ed29fee11a9a7fa/api/v1/users/PKPUserHandler.php#L142) (with a limit and offset), then the limit and offset are reset and the same query builder is [used to get the total count](https://github.com/pkp/pkp-lib/blob/d8244127bff8078f834f31f23ed29fee11a9a7fa/api/v1/users/PKPUserHandler.php#L151) with the limit and offset removed.

But this can only be done on the `main` branch, where the QueryBuilder pattern has been adopted. For `stable-3_3_0`, I'd recommend an intervention on the specific query that's causing the slowdown.

Reply:
@asmecher I just skimmed the ADODB implementation, and I think I'm able to break it yet 🙃

I was already expecting a negative answer, so I just wanted to check your opinion before (given the tests passed and there's a retry mechanism). So, I'll go with the less aggressive approach 👌

---

About the Laravel's builder, I did some tests and we can't blindly trust on its output, this simple query:
```php
Capsule::table('submissions')->groupBy('context_id')->count();
```

Outputs (the count return will be whatever comes up in the first record, doesn't fail, but completely wrong 🙈):
```sql
select count(*) as aggregate from `submissions` group by `context_id`
```

I did other tests, and not so complex constructions can easily break the output as well (I just know one near perfect query builder that also does optimizations, but it's not available for PHP).

This way, I think it's worth to review the `count()` usage! I didn't inspect alternatives yet, but looks like it will be safer for us to keep enclosing it with another `SELECT` to avoid "naive" failures (e.g. calling the `$builder->count()` on a random `Builder`).

Reply:
The issue with `GROUP BY` / `count()` is a known Laravel behaviour (see e.g. https://github.com/laravel/framework/issues/44081), and IMO it it's going to be fixed it should be fixed there. If they consider it acceptable behaviour, we should code to Laravel's expectations rather than trying to alter its behaviour. In essence, I'm trying to keep us in our lane (application code) and keep framework concerns with the framework.

Reply:
Yep. I just wanted to point out that some review might be needed at our `count()` usage later :)